### PR TITLE
QnAMaker GenerateAnswer StrictFiltersCompoundOperationType for Metadata Join Operation

### LINKED ISSUE #2719: QnAMaker GenerateAnswer filters need to support OR operation.
Is your feature request related to a problem? Please describe.
QnAMaker GenerateAnswer Filters do not support "OR" operation for Join. They currently support AND operation.
QnAMaker GenerateAnswer API has been enhanced to provide choice to user for join condition. User can choose the Compound Operation for joining strictfilters a.k.a metadata. The values for the Compound Operation Type are 'AND' and 'OR'. The default behaviour is 'AND', to support backward compatibility.

Describe the solution you'd like
QnAMaker API has introduced support for "OR" operation. The feature needs to be implemented in SDKs.

Example: {
"question":"what is metadata",
"istest": false,
"top": 3,
"strictFilters": [
{
"name": "country",
"value": "in"
},
{
"name": "country",
"value": "us"
}],
"strictFiltersCompoundOperationType": "OR"
}

Response Body:
{
"answers": [
{
"questions": [
"what is metadata"
],
"answer": "It is interface to Azure Search filters",
"score": 100.0,
"id": 1,
"source": "Editorial",
"metadata": [
{
"name": "country",
"value": "in"
}
],
"context": {
"isContextOnly": false,
"prompts": []
}
}
]
}

### PR DESCRIPTION
**Pull request by vipeketi**
Fixes #2719 
QnAMaker GenerateAnswer API has been enhanced to provide choice to user for join condition. User can choose the Compound Operation for joining strictfilters a.k.a metadata. The values for the Compound Operation Type are 'AND' and 'OR'. The default behaviour is 'AND', to support backward compatibility.

The changes implementation needs change in QnAMakerOptions, QnAMakerDialog.schema, testbot.schema and tests.schema. Changes implemented in QnARecognizer and QnAMakerDialog. Test has been created for testing the functionality.

Testing
QnaMaker_StrictFilters_Compound_OperationType() added for testing.

### CODE REVIEW

File: libraries/botbuilder-ai/src/qnamaker-interfaces/qnamakerOptions.ts [hunk 1]
```diff
@@ -64,4 +65,9 @@ export interface QnAMakerOptions {
      * Ranker types.
      */
     rankerType?: string;
+
+    /**
+     * A value indicating user's choice to for Strict Filters Join Operation. 
+     */
+    strictFiltersCompoundOperationType?: StrictFiltersCompoundOperationType;
```
Comment (somigithub) [inline libraries/botbuilder-ai/src/qnamaker-interfaces/qnamakerOptions.ts:72]: any tests covering this change?

**Review by somigithub — CHANGES_REQUESTED** [review]

File: libraries/botbuilder-ai/src/qnamaker-interfaces/qnamakerOptions.ts (hunk 1 above)
Comment (vipeketi) [reply libraries/botbuilder-ai/src/qnamaker-interfaces/qnamakerOptions.ts:72]: Yes, added two tests along with testdata to validate new field.

---
In reply to: [473518553](https://github.com/microsoft/botbuilder-js/pull/2718#discussion_r473518553) [](ancestors = 473518553)

**Review by bingisbestest — APPROVED** [review]

**Review by somigithub — APPROVED** [review]

**Review by johnataylor — APPROVED** [review]

File: libraries/botbuilder-ai/src/qnamaker-interfaces/StrictFiltersCompoundOperationType.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Enumeration of types of StrickFilters Operation Types.
+ */
+export class StrictFiltersCompoundOperationType {
```
Comment (chrimc62) [inline libraries/botbuilder-ai/src/qnamaker-interfaces/StrictFiltersCompoundOperationType.ts:12]: >StrictFiltersCompoundOperationType [](start = 13, length = 34)

As commented in the C# version, naming is horrible.

**Review by chrimc62 — CHANGES_REQUESTED** [review]
:clock1:

File: libraries/botbuilder-ai/src/qnamaker-interfaces/StrictFiltersCompoundOperationType.ts (hunk 2 above)
Comment (chrimc62) [reply libraries/botbuilder-ai/src/qnamaker-interfaces/StrictFiltersCompoundOperationType.ts:12]: You should rename these the same the C# one, i.e. JoinOperators and StrictFiltersJoinOperator

---
In reply to: [486653189](https://github.com/microsoft/botbuilder-js/pull/2718#discussion_r486653189) [](ancestors = 486653189)

File: libraries/botbuilder-ai/src/qnamaker-interfaces/StrictFiltersCompoundOperationType.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+/**
+ * @module botbuilder-ai
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Enumeration of types of StrickFilters Operation Types.
```
Comment (chrimc62) [inline libraries/botbuilder-ai/src/qnamaker-interfaces/StrictFiltersCompoundOperationType.ts]: >StrickFilters [](start = 27, length = 13)

spelling.

File: libraries/botbuilder-ai/src/qnamaker-interfaces/StrictFiltersCompoundOperationType.ts (hunk 2 above)
Comment (vipeketi) [reply libraries/botbuilder-ai/src/qnamaker-interfaces/StrictFiltersCompoundOperationType.ts]: File has been deleted for name changes, now.

File: libraries/botbuilder-ai/src/qnamaker-utils/generateAnswerUtils.ts [hunk 4]
```diff
@@ -60,20 +60,21 @@ export class GenerateAnswerUtils {
...
         const url: string = `${ endpoint.host }/knowledgebases/${ endpoint.knowledgeBaseId }/generateanswer`;
         var queryOptions: QnAMakerOptions = { ...this._options, ...options } as QnAMakerOptions;
-        
+
         queryOptions.rankerType = !queryOptions.rankerType ? RankerTypes.default : queryOptions.rankerType;
         this.validateOptions(queryOptions);
 
         var payloadBody = JSON.stringify({
             question: question,
+            strictFiltersJoinOpertionType: queryOptions.strictFiltersJoinOperator,
```
Comment (somigithub) [inline libraries/botbuilder-ai/src/qnamaker-utils/generateAnswerUtils.ts:69]: in Qna Service Json payload (left side), it should be still strictFiltersCompoundOperationType

**Pull request merged**